### PR TITLE
feat: add new package `rule-tester`

### PR DESCRIPTION
**Pull request by bradzacher**
## PR Checklist

- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

References:
- https://github.com/typescript-eslint/typescript-eslint/discussions/6499
- #5750
- #1752

## Overview

The changes I implemented in #5750 were pretty complex to build because I was working around how the base `RuleTester` class did things. I ran into a lot of weirdness due to things like the class expecting specific properties on objects.

In looking to implement #6499 - I don't want to hit the same issues, so I think it's best if we just draw a line in the sand and fork it so that we can build our enhancements directly into the tooling.

This is also good because it means that we can provide a new baseline for testing and don't need to worry about what the installed ESLint version supports - which will make our eventual dependency testing easier.

This PR:
- fork the base class and adds types
    - The logic itself is unchanged, but I did move some things around to make it a bit easier to type and manage.
- clones the base eslint test for `RuleTester`
    - This test is completely unchanged because it'll take a complete rewrite to make it jest-compatible.
- adds the changes we've added in our existing subclass fork:
    - adds `afterAll` as additional static "test framework providers" to the class.
    - adds a new per-test option - `skip` which runs the test using `it.skip`.
    - adds automatic parser cache cleaning using `afterAll`.
    - adds automatic file naming for type-aware linting
        - I implemented this slightly differently than before:
            - it uses a different default name for `tsx` files now
            - the default filenames can be overridden in the constructor (previously you HAD to do it per test).
    - adds per-test dependency constraints
        - I implemented this slightly differently than before:
          Before our hands were tied and I had to lean on setting `only` on every test to filter out the skipped tests - now with the new `skip` functionality we only need set `skip` on these tests which is a much, much nicer implementation!
- adds new features
    - adds a new per-test option - `skip` which runs the test case using `it.skip`.
    - adds `itSkip` and `describeSkip` as additional static "test framework providers" to the class.
    - adds a new `defaultFilenames` constructor option to allow configuring the default filenames used for type-aware testing.

Documentation:
https://deploy-preview-6777--typescript-eslint.netlify.app/architecture/rule-tester

TODO:
- [x] add a new package
- [x] fork the base class
- [x] implement the changes we have in [our current subclass fork](https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/utils/src/eslint-utils/rule-tester)
    - [x] implement automatic parser cache cleaning
    - [x] implement automatic file naming for type-aware linting
    - [x] implement the dependency checks from #5750
- [ ] ~~migrate codebase to this ruletester~~
    - i'll do this in a separate PR
- [ ] ~~implement snapshot testing #6499~~
    - i'll do this in a separate PR

### CODE REVIEW

File: packages/eslint-plugin-tslint/src/rules/config.ts [hunk 1]
```diff
@@ -1,10 +1,27 @@
...
+    if (cache.has(key)) {
+      return cache.get(key)!;
+    }
+    const result = func(...args);
+    cache.set(key, result as ReturnType<T>);
+    return result;
+  } as T;
+  return memoized;
+}
+
```
Comment (armano2) [inline packages/eslint-plugin-tslint/src/rules/config.ts:24]: can we extract this change to standalone PR?

File: packages/utils/src/ts-eslint/Rule.ts [hunk 2]
```diff
@@ -424,6 +425,19 @@ interface RuleListener {
...
+interface RuleListenerCatchAllBaseCase {
+  [nodeSelector: string]: RuleFunction | undefined;
+}
+// Interface to merge into for anyone that wants to add more selectors
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+interface RuleListenerExtension {}
+
+type RuleListener = RuleListenerBaseSelectors &
+  RuleListenerExitSelectors &
+  RuleListenerCatchAllBaseCase;
```
Comment (armano2) [inline packages/utils/src/ts-eslint/Rule.ts:440]: https://github.com/typescript-eslint/typescript-eslint/pull/6679/files#diff-53e6896dd0f15121e29e2dfc7309a38d6d6637fc0a5a3bfe64bc4007ab5fca64R457

maybe we could add missing codePath functions to type?

Comment (bradzacher) [reply packages/utils/src/ts-eslint/Rule.ts:440]: I just haven't bothered with them just because honestly I've never found an actual use for them in our lint rules.
Our users are apparently in the same boat because nobody has asked for them (317 dependents of our utils package listed in npm).

Maybe one day but I don't think it's worth messing up the DevX of node selectors for until we have an actual motivating usecase for.

File: packages/eslint-plugin-tslint/src/rules/config.ts (hunk 1 above)
Comment (bradzacher) [reply packages/eslint-plugin-tslint/src/rules/config.ts:24]: Yeah will do. I was just looking to see if we had a lodash dependency (because the rule tester needed one) and I noticed this plugin used the entire lodash package for just one function!

I'll extract it out because it's not worth the massive dependency, esp given we can simplify the function massively.

File: packages/utils/src/ts-eslint/Rule.ts (hunk 2 above)
Comment (armano2) [reply packages/utils/src/ts-eslint/Rule.ts:440]: https://github.com/SonarSource/eslint-plugin-sonarjs/blob/4dfd992eeccafca92caed52889b4f83bb1e38c3c/src/rules/no-one-iteration-loop.ts#L41

one of examples :P

File: packages/eslint-plugin-tslint/src/rules/config.ts (hunk 1 above)
Comment (armano2) [reply packages/eslint-plugin-tslint/src/rules/config.ts:24]: if possible we should just drop it

Comment (bradzacher) [reply packages/eslint-plugin-tslint/src/rules/config.ts:24]: https://github.com/typescript-eslint/typescript-eslint/pull/6810

File: packages/rule-tester/typings/eslint.d.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      // ...
+    };
+    environments: Map<string, Linter.Environment>;
+    // ...
+  };
+}
+
+declare module 'eslint' {
+  export { SourceCode } from '@typescript-eslint/utils/ts-eslint';
+}
```
Comment (JoshuaKGoldberg) [inline packages/rule-tester/typings/eslint.d.ts:24]: Lol.

**Review by JoshuaKGoldberg — APPROVED** [review]
Whew, what a PR! This is great - looking forward to having native dependency contraints!

File: docs/architecture/Rule_Tester.mdx [hunk 4]
```diff
@@ -0,0 +1,216 @@
+---
+id: rule-tester
+sidebar_label: rule-tester
+---
+
+import CodeBlock from '@theme/CodeBlock';
+
+# `@typescript-eslint/rule-tester`
```
[the remark this answers is not in the document]
Comment (bradzacher) [reply docs/architecture/Rule_Tester.mdx:8]: As per our chats on discord - @armano2 and I think we should push to make this a breaking change replacement in v6 - people should switch to it cos it's better in many ways and it should mostly be a 1:1 replacement.

Also switching to the package means people get the `only: boolean` test prop for free, regardless of their ESLint version.

File: docs/architecture/Rule_Tester.mdx [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+With that config the parser will automatically run in type-aware mode and you can write tests just like before.
+
+### Test Dependency Constraints
+
+Sometimes it's desirable to test your rule against multiple versions of a dependency to ensure backwards and forwards compatibility.
+With backwards-compatibility testing there comes a complication in that some tests may not be compatible with an older version of a dependency.
+For example - if you're testing against an older version of TypeScript, certain features might cause a parser error!
+
+import DependencyConstraint from '!!raw-loader!../../packages/rule-tester/src/types/DependencyConstraint.ts';
```
[the remark this answers is not in the document]
Comment (bradzacher) [reply docs/architecture/Rule_Tester.mdx:155]: It's much nicer than us copy-pasting and manually syncing things fo sho

File: packages/rule-tester/src/RuleTester.ts [hunk 6]
```diff
@@ -0,0 +1,1028 @@
...
+  AnyRuleModule,
+  ParserOptions,
+  RuleContext,
+  RuleModule,
+} from '@typescript-eslint/utils/ts-eslint';
+import { Linter } from '@typescript-eslint/utils/ts-eslint';
+// we intentionally import from eslint here because we need to use the same class
+// that ESLint uses, not our custom override typed version
+import { SourceCode } from 'eslint';
+import merge from 'lodash.merge';
```
[the remark this answers is not in the document]
Comment (bradzacher) [reply packages/rule-tester/src/RuleTester.ts:21]: I thought about that - but ESLint has this exact import for its rule tester.
So I figured that we'd just use the exact same code to avoid a duplicate install

File: packages/utils/src/ts-eslint/Rule.ts [hunk 7]
```diff
@@ -424,6 +425,18 @@ interface RuleListener {
...
   YieldExpression?: RuleFunction<TSESTree.YieldExpression>;
 }
+type RuleListenerExitSelectors = {
+  [K in keyof RuleListenerBaseSelectors as `${K}:exit`]: RuleListenerBaseSelectors[K];
+};
+interface RuleListenerCatchAllBaseCase {
+  [nodeSelector: string]: RuleFunction | undefined;
+}
+// Interface to merge into for anyone that wants to add more selectors
+interface RuleListenerExtension {}
```
[the remark this answers is not in the document]
Comment (bradzacher) [reply packages/utils/src/ts-eslint/Rule.ts:435]: That was just me adding some future-proofing in case anyone was doing something weird.
Actually don't need it - but I think it's fine to leave it undocumented for the moment.

File: packages/utils/src/json-schema.ts [hunk 8]
```diff
@@ -1,24 +1,2 @@
-// Note - @types/json-schema@7.0.4 added some function declarations to the type package
-// If we do export *, then it will also export these function declarations.
-// This will cause typescript to not scrub the require from the build, breaking anyone who doesn't have it as a dependency
```
[the remark this answers is not in the document]
Comment (bradzacher) [reply packages/utils/src/json-schema.ts:3]: yeah because in the olden days we did `export * from 'json-schema'` and we'd export the types for functions that don't actually exist.
But now a days we can do `export type *` and it'll not include the runtime bits!

Though #6963 would just be better than making this change at all TBH
